### ISSUE TITLE
inconsistent student project restrictions configuration -- add a save button and state

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#5310]
~~I enabled some restrictions to develop those ssh and internet access blocks. Problem is, clicking those checkboxes triggers an immediate action, which somehow results in inconsistent values. Below is what ended up in the DB. Closing/opening the course editor did trigger a reconfiguration of all student projects, but the that inconsistent data stayed as it is.~~

This was a bit confusing to me. The goal of this ticket is to add an explicit save button to confirm the changes. ... in particular, doing two or more concurrent rounds of setting the configuration after two or more checkbox clicks should be avoided.

```
smc=# select project_id, course -> 'student_project_functionality' from projects where course ->> 'project_id' = '....';
              project_id              |                   ?column?
--------------------------------------+----------------------------------------------
 831b4b0e-b1b4-483f-b41b-25a17745d8e3 |
 fccda67e-7c58-4de8-a7ab-d32c936217fd |
 a07ddbf8-a192-4d51-a9da-f1f1c7ba3af4 |
 73a532dd-3ccc-4b1c-88d0-50c2565e2e9f |
 bc277848-aa91-49b8-ad96-e026907825fd |
 a5b9acbe-0e2b-432c-87c0-09d15c1ad661 | {"disableSSH": true, "disableNetwork": true}
 235e51c6-969a-43d5-9ed1-339dce094900 |
 571ab5fd-373a-4974-bcda-83816c9e05d3 |
 f1ad89fc-7572-4d0a-8b2e-ed39f925a228 |
 7ee84013-a78c-4fd5-9e31-8cb7d1c77c38 | {"disableSSH": true, "disableNetwork": true}
 29bf687b-a428-499d-8f35-98e39aaa8d46 | {"disableSSH": true, "disableNetwork": true}
 25575228-e81f-49c3-bd60-0950dae11d4e | {"disableSSH": true, "disableNetwork": true}
 5a13b2a0-41bc-488a-a1c7-64fbc0271e21 | {"disableSSH": true, "disableNetwork": true}
 1777dc61-a550-4d13-b1ab-9739cd2348f7 | {"disableSSH": true, "disableNetwork": true}
 4c45f74b-1a35-4453-b783-e53d9670b17a | {"disableSSH": true, "disableNetwork": true}
[...]
```

I think an explicit "Save" button would be better, as well as disabling it while the configuration is active. 

~~I also wonder why all projects need to be configured. Wouldn't it be enough to store the field in the teacher project's course field? Or is the idea to make this a per-student-project setting?  .... Oh, I forgot, there is nothing for the teacher project, there could be multiple courses. OK ok...~~

**Comment by haraldschilly** [comment]
Ok, maybe this is not a bug. I'm still puzzled why some aren't configured. There are no deleted students, at least none are visible in the UI. But there are more (43) than students (35). Very strange. I'll close this ticket for now.

**Comment by williamstein** [comment]
> I think an explicit "Save" button would be better, as well as disabling it while the configuration is active.

Yes, that would be good to implement.

> I also wonder why all projects need to be configured. Wouldn't it be enough to store the field in the teacher project's course field? 

For the backend disableSSH and disableNetwork, that would I guess be technically possible.  For all other settings, it is critical that the setting be in all projects, since they are implemented **on the frontend** by the student's browser, and the student doesn't have any access whatever to the instructor project.

**Comment by williamstein** [comment]
> explicit save button

Let's re-open this issue with the title "explicit save button".  When there was only one checkbox, it didn't make any difference, but not that there are a bunch of them, it is not very usable without some additional UI.  E.g., (1) an explicit Save, and (2) while that configuration happens the UI should be disabled so you can't save additional changes.

**Comment by haraldschilly** [comment]
yes, ok

**Comment by williamstein** [comment]
OK, I implemented this in ee80a8fd369610f0148176628a0343a497102908.  It was easy to do using react hooks, but also adds a lot more code and complexity, so ugh.  It should be less confusing.
